src/mappings/entities/mapTickerExternalAgentHistory.ts

```diff
@@ -0,0 +1,291 @@
+          datetime: event.block.timestamp,
+          type: "AgentPermissionsChanged",
+          permissions,
+        }).save()
+      );
+    }
+    await Promise.all(promises);
+    return;
+  }
+
+  if (
+    moduleId === ModuleIdEnum.Externalagents &&
+    eventId === EventIdEnum.AgentAdded
+  ) {
+    const did = params[0].toString();
+    const group = params[2].toJSON() as AgentGroup;
+    const ticker = serializeTicker(params[1]);
+
+    const promises = [
+      (async () => {
+        const permissions = await permissionsFromAgentGroup(
+          ticker,
+          group,
+          async (n) => {
+            const ag = await AgentGroupEntity.get(`${ticker}/${n}`);
+            return ag.permissions;
+          }
+        );
+        await TickerExternalAgentHistory.create({
+          id: `${blockId}/${eventIdx}/${did}`,
+          ticker,
+          did,
+          blockId,
+          eventIdx,
+          datetime: event.block.timestamp,
+          type: "AgentAdded",
+          permissions,
+        }).save();
+      })(),
+    ];
+
+    // Only keep track of membership for custom agent groups.
+    if (isCustom(group)) {
+      promises.push(
+        AgentGroupMembership.create({
+          id: `${ticker}/${group.custom}/${did}`,
+          member: did,
+          group: `${ticker}/${group.custom}`,
+        }).save()
+      );
+    }
+    await Promise.all(promises);
+    return;
+  }
+
+  if (
+    moduleId === ModuleIdEnum.Externalagents &&
+    eventId === EventIdEnum.GroupChanged
+  ) {
+    const did = params[2].toString();
+    const group = params[3].toJSON() as AgentGroup;
+    const ticker = serializeTicker(params[1]);
+
+    const promises = [
+      removeMember(did, ticker),
+      (async () => {
+        const permissions = await permissionsFromAgentGroup(
+          ticker,
+          group,
+          async (n) => {
+            const ag = await AgentGroupEntity.get(`${ticker}/${n}`);
+            return ag.permissions;
+          }
+        );
+        await TickerExternalAgentHistory.create({
+          id: `${blockId}/${eventIdx}/${did}`,
+          ticker,
+          did,
+          blockId,
+          eventIdx,
+          datetime: event.block.timestamp,
+          type: "AgentPermissionsChanged",
+          permissions,
+        }).save();
+      })(),
+    ];
+
+    // Only keep track of membership for custom agent groups.
+    if (isCustom(group)) {
+      promises.push(
+        AgentGroupMembership.create({
+          id: `${ticker}/${group.custom}/${did}`,
+          member: did,
+          group: `${ticker}/${group.custom}`,
+        }).save()
+      );
+    }
+    await Promise.all(promises);
+    return;
+  }
+
+  if (
+    moduleId === ModuleIdEnum.Externalagents &&
+    eventId === EventIdEnum.AgentRemoved
+  ) {
+    const did = params[2].toString();
+    const ticker = serializeTicker(params[1]);
+
+    const promises = [
+      removeMember(did, ticker),
+      (async () => {
```

Comment:
(nit/ignorable) I don't think we need an anonymous function here since nothing happens after await.